crates/dbsp/src/circuit/circuit_builder.rs

```diff
@@ -4935,22 +5843,357 @@ impl CircuitHandle {
+    /// 2. Identify stateful operators (such as integrals and output nodes) that don't have
+    ///    a checkpoint and require backfill (the `need_backfill` set).
+    /// 3. Iterate backward from `need_backfill` nodes to find all operators that
+    ///    should participate in the replay phase of the circuit. Iteration stops when
+    ///    reaching a stream whose contents can be replayed from an existing node
+    ///    that has a checkpoint or an input node.
+    /// 4. If the circuit requires backfill, prepare the circuit for replay by
+    ///    configuring the scheduler to only schedule nodes that participate in
+    ///    backfill.
+    ///
+    /// Returns `None` if the circuit does not require backfill; returns info about
+    /// nodes that participate in backfill otherwise.
+    ///
+    /// * After calling this function, the client can invoke `step` repeatedly for replay to make progress.
+    /// * Use `is_replay_complete` to determine whether the circuit has finished the replay.
+    /// * Use `complete_replay` to finalize the replay phase and prepare the circuit for normal operation.
+    pub fn restore(&mut self, base: &StoragePath) -> Result<Option<BootstrapInfo>, DbspError> {
+        // Nodes that will act as replay sources during the replay phase of the circuit.
+        let mut replay_sources: BTreeMap<NodeId, StreamId> = BTreeMap::new();
+
+        // Nodes that require backfill from upstream nodes.
+        let mut need_backfill: BTreeSet<GlobalNodeId> = BTreeSet::new();
+
+        // debug!("CircuitHandle::restore: restoring from checkpoint {}", base);
+
+        // Initialize `need_backfill` to operators without a checkpoint.
+        // Fail if there are any errors other than NotFound.
+        // By the end of this, `need_backfill` will contain all new integrals and output
+        // nodes that need backfill.
+        self.circuit.map_nodes_recursive_mut(
+            &mut |node: &mut dyn Node| match node.restore(base) {
+                Err(e) if Runtime::mode() == Mode::Ephemeral => Err(e),
+                Err(DbspError::Storage(ioerror)) if ioerror.kind() == ErrorKind::NotFound => {
+                    need_backfill.insert(node.global_id().clone());
+                    Ok(())
+                }
+                Err(DbspError::IO(ioerror)) if ioerror.kind() == ErrorKind::NotFound => {
+                    need_backfill.insert(node.global_id().clone());
+                    Ok(())
+                }
+                Err(e) => Err(e),
+                Ok(()) => Ok(()),
+            },
+        )?;
+
+        debug!(
+            "worker {}: CircuitHandle::restore: found {} operators that require backfill: {:?}",
+            Runtime::worker_index(),
+            need_backfill.len(),
+            need_backfill.iter().cloned().collect::<Vec<GlobalNodeId>>()
+        );
+
+        // We can only backfill a nested circuit as a whole, so if we encounter at least
```

Comment:
I guess this replies my question
also, there can be no inputs in nested circuits